### PR TITLE
Const fold create list

### PR DESCRIPTION
**Pull request by pjmore**
Adds optimization for `list_index(list_create, literal)` to MirScalarExpr::reduce. The optimization inlines the list_index function when the index is a literal integer. If the integer is an invalid index for the list the expression is unchanged.

### Motivation

* This PR adds a known-desirable feature: #9020

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: src/expr/src/scalar/mod.rs [hunk 1]
```diff
@@ -632,6 +632,23 @@ impl MirScalarExpr {
...
                                 ),
                             };
+                        } else if *func == VariadicFunc::ListIndex && exprs[1].is_literal_ok() {
+                            let literal = match exprs[1].as_literal().unwrap().unwrap() {
+                                Datum::Int64(data) => usize::try_from(data - 1).ok(),
+                                _ => None,
+                            };
+                            if let Some(index_literal) = literal {
+                                if let MirScalarExpr::CallVariadic {
+                                    func,
```
Comment (wangandi) [inline src/expr/src/scalar/mod.rs:642]: You should be able to change this line to `func: VariadicFunc::ListCreate {..},` and skip the matching for `VariadicFunc::ListCreate { .. }` further down.

File: src/expr/src/scalar/mod.rs [hunk 2]
```diff
@@ -1557,6 +1574,64 @@ mod tests {
...
+                    ]
+                },
+                output: MirScalarExpr::CallVariadic{
+                    func: VariadicFunc::ListIndex,
+                    exprs: vec![
+                        MirScalarExpr::CallVariadic{
+                            func: VariadicFunc::ListCreate{elem_type: ScalarType::Int64},
+                            exprs: vec![col(0), col(1), col(2)]
+                        },
+                        lit(10)
```
Comment (wangandi) [inline src/expr/src/scalar/mod.rs:1630]: Besides a test for multiple layers of nested lists, add a test for:
* `index = 0`, 
* `index = <list length + 1>`

Also, try varying the length of lists created and the expressions involved.

**Review by wangandi — COMMENTED** [review]
Thank you for taking on this issue!

This current PR is missing support for a couple of cases. If you look at the implementation for the function `list_index`
https://github.com/MaterializeInc/materialize/blob/67e8980f578c1903e01b184982673f65657de1d3/src/expr/src/scalar/func.rs#L5182
You will see that:
* `list_index` is supposed to return `null` if the index is 1) less than 1 or 2) greater than the length of the list.
* `list_index` is a variadic function that can index into multiple layers of lists. 

For example,
```
materialize=> explain select list[list[f1, f2], list[f3, f4]][2][1] from t1;
                                Optimized Plan                                 
-------------------------------------------------------------------------------
 %0 =                                                                         +
 | Get materialize.public.t1 (u26)                                            +
 | Map list_index(list_create(list_create(#0, #1), list_create(#2, #3)), 2, 1)+
 | Project (#4)                                                               +
```
should become:
```
 %0 =                                                                         +
 | Get materialize.public.t1 (u26)                                            +
 | Project (#2)
```

Also, please run `cargo fmt` so that your PR passes the formatting check.

**Comment by pjmore** [comment]
@wangandi Thanks for the feedback, I didn't consider multi-dimensional lists at all. I'm going to close this till I've fixed the issues you mentioned.

**Pull request closed without merging**